

## Введение в Pandas 

Pandas - библиотека для работы с табличными данными в питоне.

Чтобы не писать название библиотеки целиком каждый раз, когда понадобится ее использовать, принято сокращать название библиотеки и импортировать ее как "pd":

In [1]:
import pandas as pd # импортировали библиотеку pandas и назвали ее pd 

В Pandas есть тип данных датафрейм (DataFrame), в котором удобно хранить таблицы с данными. Создадим объект такого типа самостоятельно:

In [2]:
df = pd.DataFrame() # создали пустой датафрейм с помощью метода DataFrame() библиотеки pandas (pd)
df['a'] = [10,20,30] # создаем колонку "а" и помещаем в нее столбец с данными - [10, 20, 30]
df

a
0  10
1  20
2  30

In [3]:
df['b'] = ['one', 'two', 'three']
df

a      b
0  10    one
1  20    two
2  30  three

Конечно, чаще всего приходится работать с уже готовыми наборами данных. Такие данные обычно хранятся в формтае xls(x) - для работы в Excel, или (чаще) в формате csv - comma-separated value. Попробуем импортировать csv файл с данными о пассажирах Титаника.

In [4]:
data = pd.read_csv('https://raw.githubusercontent.com/rogovich/Data/master/data/titanic/train.csv')

Функция read_csv читает данные из файла формата csv и преобразует в pandas.DataFrame. Аналогичная функция read_excel может читать данные в офрмате xls(x).

Посмотрим на наши данные:

In [5]:
data.head() # функция head() показывает первые строки датафрейма, по умолчанию 5

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [6]:
data.head(10) # можно передать аргументом количество строк, которые хотите увидеть

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [7]:
data.tail(10) # можно посмотреть последние записи

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

По столбцам идут признаки, по строкам - объекты (пассажиры).

In [8]:
data.shape # функция shape показывает размерность датафрейма (строк, столбцов)

(891, 12)

In [9]:
data.columns # список столбцов 

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Описание признаков:

**PassengerId** - id пассажира

**Survived** бинарная переменная: выжил пассажирил (1) или нет (0)

**Pclass** - класс пассажира

**Name** - имя пассажира

**Sex** - пол пассажира

**Age** - возраст пассажира

**SibSp** - количество родственников (братьев, сестер, супругов) пассажира на борту

**Parch** - количество родственников (родителей / детей) пассажира на борту

**Ticket** - номер билета

**Fare** - тариф (стоимость билета)

**Cabin** - номер кабины

**Embarked** - порт, в котором пассажир сел на борт (C - Cherbourg, S - Southampton, Q = Queenstown)

Так можно обратиться к столбцу:

In [10]:
data['Age'].head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

In [11]:
data.Age.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

Или к нескольким столбцам сразу:

In [12]:
data[['Age', 'Sex']].head()

Age     Sex
0  22.0    male
1  38.0  female
2  26.0  female
3  35.0  female
4  35.0    male

A так - к строке по индексу:

In [13]:
data.loc[0]

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

In [14]:
data.iloc[0]

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

In [15]:
data.loc[1:3] # строки с 1 по 3

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

In [16]:
data.loc[1:3, 'Survived':'Sex'] # строки с 1 по 3, колонки от Survived до Sex

Survived  Pclass                                               Name     Sex
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female
2         1       3                             Heikkinen, Miss. Laina  female
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female

__loc__ возвращает данные на основе индекса, а __iloc__ -  основываясь исключительно на позиции индекса, начиная с 0.

Пример:

In [17]:
df = pd.DataFrame()
df['Name'] =  ['Tom', 'Jack', 'Nick', 'Juli']
df['Mark'] = [99, 98, 95, 90]
df.index = [1,3,2,0]
df

Name  Mark
1   Tom    99
3  Jack    98
2  Nick    95
0  Juli    90

In [18]:
df.loc[1]

Name    Tom
Mark     99
Name: 1, dtype: object

In [19]:
df.iloc[1]

Name    Jack
Mark      98
Name: 3, dtype: object

In [20]:
df.loc[1:2]

Name  Mark
1   Tom    99
3  Jack    98
2  Nick    95

In [21]:
df.iloc[1:2]

Name  Mark
3  Jack    98

Можем сделать индексом даже колоноку с именем. И тоже сможем обращаться к объектам через .loc

In [22]:
df.index = df.Name
df.loc['Nick']['Mark']

95

In [23]:
df.loc['Jack':'Nick']

Name  Mark
Name            
Jack  Jack    98
Nick  Nick    95

Кроме того, можно выбирать объекты, удовлетворяющие каким-то свойствам, например, все пассажиры-женщины:

In [24]:
data[data.Sex == 'female'].head()

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Пассажиры первого класса:

In [25]:
data[data.Pclass == 1].head()

PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
6             7         0       1   
11           12         1       1   
23           24         1       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
3       0    113803  53.1000  C123        S  
6       0     17463  51.8625   E46        S  
11      0    113783  26.5500  C103        S  
23      0    113788  35.5000    A6        S

Пассажиры первого или второго классов:

In [26]:
data[data.Pclass.isin([1,2])].head()

PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
6             7         0       1   
9            10         1       2   
11           12         1       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
3       0    113803  53.1000  C123        S  
6       0     17463  51.8625   E46        S  
9       0    237736  30.0708   NaN        C  
11      0    113783  26.5500  C103        S

Пассажиры младше 18:

In [27]:
data[data.Age < 18].head()

PassengerId  Survived  Pclass                                  Name  \
7             8         0       3        Palsson, Master. Gosta Leonard   
9            10         1       2   Nasser, Mrs. Nicholas (Adele Achem)   
10           11         1       3       Sandstrom, Miss. Marguerite Rut   
14           15         0       3  Vestrom, Miss. Hulda Amanda Adolfina   
16           17         0       3                  Rice, Master. Eugene   

       Sex   Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
7     male   2.0      3      1   349909  21.0750   NaN        S  
9   female  14.0      1      0   237736  30.0708   NaN        C  
10  female   4.0      1      1  PP 9549  16.7000    G6        S  
14  female  14.0      0      0   350406   7.8542   NaN        S  
16    male   2.0      4      1   382652  29.1250   NaN        Q

Девушки в возрасте от 18 до 25, путешествующие в одиночку (без каких-либо родственников):

In [28]:
data[(data.Sex == 'female') & (data.Age > 18) & (data.Age < 25) & (data.SibSp == 0) &(data.Parch == 0)]

PassengerId  Survived  Pclass  \
44            45         1       3   
56            57         1       2   
106          107         1       3   
141          142         1       3   
199          200         0       2   
289          290         1       3   
293          294         0       3   
310          311         1       1   
345          346         1       2   
369          370         1       1   
376          377         1       3   
404          405         0       3   
427          428         1       2   
473          474         1       2   
474          475         0       3   
501          502         0       3   
554          555         1       3   
627          628         1       1   
641          642         1       1   
649          650         1       3   
708          709         1       1   
710          711         1       1   
816          817         0       3   
882          883         0       3   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
44                       Devaney, Miss. Margaret Delia  female  19.0      0   
56                                   Rugg, Miss. Emily  female  21.0      0   
106                   Salkjelsvik, Miss. Anna Kristine  female  21.0      0   
141                           Nysten, Miss. Anna Sofia  female  22.0      0   
199             Yrois, Miss. Henriette ("Mrs Harbeck")  female  24.0      0   
289                               Connolly, Miss. Kate  female  22.0      0   
293                                Haas, Miss. Aloisia  female  24.0      0   
310                     Hays, Miss. Margaret Bechstein  female  24.0      0   
345                      Brown, Miss. Amelia "Mildred"  female  24.0      0   
369                      Aubart, Mme. Leontine Pauline  female  24.0      0   
376                    Landergren, Miss. Aurora Adelia  female  22.0      0   
404                            Oreskovic, Miss. Marija  female  20.0      0   
427  Phillips, Miss. Kate Florence ("Mrs Kate Louis...  female  19.0      0   
473       Jerwan, Mrs. Amin S (Marie Marthe Thuillard)  female  23.0      0   
474                        Strandberg, Miss. Ida Sofia  female  22.0      0   
501                                Canavan, Miss. Mary  female  21.0      0   
554                                 Ohman, Miss. Velin  female  22.0      0   
627                      Longley, Miss. Gretchen Fiske  female  21.0      0   
641                               Sagesser, Mlle. Emma  female  24.0      0   
649                    Stanley, Miss. Amy Zillah Elsie  female  23.0      0   
708                               Cleaver, Miss. Alice  female  22.0      0   
710   Mayne, Mlle. Berthe Antonine ("Mrs de Villiers")  female  24.0      0   
816                       Heininen, Miss. Wendla Maria  female  23.0      0   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   

     Parch            Ticket      Fare Cabin Embarked  
44       0            330958    7.8792   NaN        Q  
56       0        C.A. 31026   10.5000   NaN        S  
106      0            343120    7.6500   NaN        S  
141      0            347081    7.7500   NaN        S  
199      0            248747   13.0000   NaN        S  
289      0            370373    7.7500   NaN        Q  
293      0            349236    8.8500   NaN        S  
310      0             11767   83.1583   C54        C  
345      0            248733   13.0000   F33        S  
369      0          PC 17477   69.3000   B35        C  
376      0            C 7077    7.2500   NaN        S  
404      0            315096    8.6625   NaN        S  
427      0            250655   26.0000   NaN        S  
473      0   SC/AH Basle 541   13.7917     D        C  
474      0              7553    9.8375   NaN        S  
501      0            364846    7.7500   NaN        Q  
554      

Найдем женщин от 18 до 25 без родственников. Сколько таких путещественниц?

In [29]:
data[(data.Sex == 'female') & (data.Age > 18) & (data.Age < 25) & (data.SibSp == 0) &(data.Parch == 0)].shape

(25, 12)

Иногда нужно создать новый признак из уже существующих, например, нам интересно, сколько всего родственников путешествовало с каждым пассажиром - просто сложим столбцы SibSp и Parch и поместим сумму в новый столбец FamilySize.

In [30]:
data['FamilySize'] = data['SibSp'] + data['Parch']
data.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  FamilySize  
0      0         A/5 21171   7.2500   NaN        S           1  
1      0          PC 17599  71.2833   C85        C           1  
2      0  STON/O2. 3101282   7.9250   NaN        S           0  
3      0            113803  53.1000  C123        S           1  
4      0            373450   8.0500   NaN        S           0  
5      0            330877   8.4583   NaN        Q           0  
6      0             17463  51.8625   E46        S           0  
7      1            349909  21.0750   NaN        S           4  
8      2            347742  11.1333   NaN        S           2  
9      0            237736  30.0708   NaN        C           1

А теперь давайте создадим переменную, которая бы нам показывала, что пассажир ехал в одиночку. Такой пассажир путешествовал без родственников. Мы напишем условие с помощью анонимной функции (1, если FamilySize равно 0 и 0 во всех остальных случаях) и применим ее к столбцу FamilySize с помощью метода .apply().

In [31]:
data['Alone'] = data['FamilySize'].apply(lambda x: 1 if x == 0 else 0)
data.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  FamilySize  Alone  
0      0         A/5 21171   7.2500   NaN        S           1      0  
1      0          PC 17599  71.2833   C85        C           1      0  
2      0  STON/O2. 3101282   7.9250   NaN        S           0      1  
3      0            113803  53.1000  C123        S           1      0  
4      0            373450   8.0500   NaN        S           0      1  
5      0            330877   8.4583   NaN        Q           0      1  
6      0             17463  51.8625   E46        S           0      1  
7      1            349909  21.0750   NaN        S           4      0  
8      2            347742  11.1333   NaN        S           2      0  
9      0            237736  30.0708   NaN        C           1      0

Функция, которую применяем к столбцу, может быть и посложнее. Давайте из каждого имени пассажира достанем его титул. Сначала потренируемся на одном пассажире.

In [32]:
data.loc[0]['Name']

'Braund, Mr. Owen Harris'

Ок, выбрали имя. Это строка. Давайте подумаем, как достать из нее титул. Вроде бы титул всегда после фамилии. Разобьем строку по запятой и достанем второй (первый по индексу) элемент. Потом разобьем по точке и возьмем первый (нулевой) элемент. И избавимся от пробела.

In [33]:
data.loc[0]['Name'].split(',')[1].split('.')[0].strip()

'Mr'

Ура! Теперь напишем функцию, которая будет забирать титул из имени, а потом применим ее к колонке Name.

In [34]:
def return_title(full_name):
    return(full_name.split(',')[1].split('.')[0].strip())

Теперь сформируем новый столбец family_name из столбца Name с помощью написанной нами функции:

In [35]:
data['Title'] = data.Name.apply(return_title)
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FamilySize  Alone Title  
0      0         A/5 21171   7.2500   NaN        S           1      0    Mr  
1      0          PC 17599  71.2833   C85        C           1      0   Mrs  
2      0  STON/O2. 3101282   7.9250   NaN        S           0      1  Miss  
3      0            113803  53.1000  C123        S           1      0   Mrs  
4      0            373450   8.0500   NaN        S           0      1    Mr

Полезно посмотреть, какие значения содержатся в столбце. Для категориальных данных можно посчитать частоту встречаемости каждого значения с помошью функции value_counts:

In [36]:
data['Title'].value_counts()

Mr              517
Miss            182
Mrs             125
Master           40
Dr                7
Rev               6
Mlle              2
Major             2
Col               2
the Countess      1
Capt              1
Ms                1
Sir               1
Lady              1
Mme               1
Don               1
Jonkheer          1
Name: Title, dtype: int64

Очень много уникальных значений! Обычно это не очень хорошо для статистического анализа, давайте все такие титулы переиминуем в Misc (другие).

In [37]:
data['Title'] = data['Title'].apply(lambda x: 'Misc' if x not in ['Mr', 'Miss', 'Mrs', 'Master'] else x)

In [38]:
data['Title'].value_counts()

Mr        517
Miss      182
Mrs       125
Master     40
Misc       27
Name: Title, dtype: int64

In [39]:
data['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [40]:
data['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

Для количественных данных удобнее смотреть минимальные/максимальные/средние значения:

In [41]:
print(data['Age'].min()) 
print(data['Age'].max()) 
print(data['Age'].mean())

0.42
80.0
29.69911764705882


В Pandas есть функция describe(), которая делает удобную сводную таблицу по всем количественным столбцам сразу (обратите внимание, что для Pandas количественные данные = все, что представлено числами, что, разумеется, неверно в общем случае):

In [42]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  FamilySize       Alone  
count  891.000000  891.000000  891.000000  891.000000  
mean     0.381594   32.204208    0.904602    0.602694  
std      0.806057   49.693429    1.613459    0.489615  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    7.910400    0.000000    0.000000  
50%      0.000000   14.454200    0.000000    1.000000  
75%      0.000000   31.000000    1.000000    1.000000  
max      6.000000  512.329200   10.000000    1.000000

In [43]:
data[['Age', 'Fare']].describe() # также можно применять только к отдельным колонкам

Age        Fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

Данные можно сортировать:

In [44]:
data = data.dropna(how = 'any')

In [45]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 183 entries, 1 to 889
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  183 non-null    int64  
 1   Survived     183 non-null    int64  
 2   Pclass       183 non-null    int64  
 3   Name         183 non-null    object 
 4   Sex          183 non-null    object 
 5   Age          183 non-null    float64
 6   SibSp        183 non-null    int64  
 7   Parch        183 non-null    int64  
 8   Ticket       183 non-null    object 
 9   Fare         183 non-null    float64
 10  Cabin        183 non-null    object 
 11  Embarked     183 non-null    object 
 12  FamilySize   183 non-null    int64  
 13  Alone        183 non-null    int64  
 14  Title        183 non-null    object 
dtypes: float64(2), int64(7), object(6)
memory usage: 22.9+ KB


In [46]:
data.sort_values(by=['Age']).tail() # сортируем по возрасту, по умолчанию сортирвка по возрастанию

PassengerId  Survived  Pclass                                  Name  \
54            55         0       1        Ostby, Mr. Engelhart Cornelius   
456          457         0       1             Millet, Mr. Francis Davis   
745          746         0       1          Crosby, Capt. Edward Gifford   
96            97         0       1             Goldschmidt, Mr. George B   
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   

      Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  FamilySize  \
54   male  65.0      0      1     113509  61.9792   B30        C           1   
456  male  65.0      0      0      13509  26.5500   E38        S           0   
745  male  70.0      1      1  WE/P 5735  71.0000   B22        S           2   
96   male  71.0      0      0   PC 17754  34.6542    A5        C           0   
630  male  80.0      0      0      27042  30.0000   A23        S           0   

     Alone Title  
54       0    Mr  
456      1    Mr  
745      0  Misc  
96       1    Mr  
630      1    Mr

In [47]:
data.sort_values(by=['Age'], ascending=False).head() # сортируем по возрасту, теперь по убыванию

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   
96            97         0       1             Goldschmidt, Mr. George B   
745          746         0       1          Crosby, Capt. Edward Gifford   
456          457         0       1             Millet, Mr. Francis Davis   
54            55         0       1        Ostby, Mr. Engelhart Cornelius   

      Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  FamilySize  \
630  male  80.0      0      0      27042  30.0000   A23        S           0   
96   male  71.0      0      0   PC 17754  34.6542    A5        C           0   
745  male  70.0      1      1  WE/P 5735  71.0000   B22        S           2   
456  male  65.0      0      0      13509  26.5500   E38        S           0   
54   male  65.0      0      1     113509  61.9792   B30        C           1   

     Alone Title  
630      1    Mr  
96       1    Mr  
745      0  Misc  
456      1    Mr  
54       0    Mr

In [48]:
data.sort_values(by=['Age', 'Fare'], ascending=False).head() # сортируем сперва по возрасту (по убыванию),
                                                             # потом стоимости билета  (по убыванию)

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   
96            97         0       1             Goldschmidt, Mr. George B   
745          746         0       1          Crosby, Capt. Edward Gifford   
54            55         0       1        Ostby, Mr. Engelhart Cornelius   
456          457         0       1             Millet, Mr. Francis Davis   

      Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  FamilySize  \
630  male  80.0      0      0      27042  30.0000   A23        S           0   
96   male  71.0      0      0   PC 17754  34.6542    A5        C           0   
745  male  70.0      1      1  WE/P 5735  71.0000   B22        S           2   
54   male  65.0      0      1     113509  61.9792   B30        C           1   
456  male  65.0      0      0      13509  26.5500   E38        S           0   

     Alone Title  
630      1    Mr  
96       1    Mr  
745      0  Misc  
54       0    Mr  
456      1    Mr

In [49]:
data.sort_values(by=['Age', 'Fare'], ascending=[False, True]).head() # сортируем сперва по возрасту (по убыванию),
                                                                     # потом стоимости билета  (по возрастанию)

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   
96            97         0       1             Goldschmidt, Mr. George B   
745          746         0       1          Crosby, Capt. Edward Gifford   
456          457         0       1             Millet, Mr. Francis Davis   
54            55         0       1        Ostby, Mr. Engelhart Cornelius   

      Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  FamilySize  \
630  male  80.0      0      0      27042  30.0000   A23        S           0   
96   male  71.0      0      0   PC 17754  34.6542    A5        C           0   
745  male  70.0      1      1  WE/P 5735  71.0000   B22        S           2   
456  male  65.0      0      0      13509  26.5500   E38        S           0   
54   male  65.0      0      1     113509  61.9792   B30        C           1   

     Alone Title  
630      1    Mr  
96       1    Mr  
745      0  Misc  
456      1    Mr  
54       0    Mr

Сохранение датафрейма:

In [50]:
data.to_csv('new_titanic.csv')

## Описательные статистики в Python


**Максимальное и минимальное значения**

In [51]:
data.Age.min()

0.92

In [52]:
data.Age.max()

80.0

**Меры центральной тенденции**

Меры среднего уровня дают усредненную характеристику совокупности объектов по определенному признаку.

В зависимости от типа шкалы измерения применяются следующие меры среднего уровня:
- для количественных данных - среднее значение (арифметическое среднее), медиана, мода

- для категориальных:

   для порядковых данных - медиана, мода

   для номинальных данных - мода

На самом деле таких мер довольно много, например: взвешенное среднее — среднее значение, учитывающее весовые коэффициенты для каждого значения, гармоническое среднее — количество наблюдений, деленное на сумму инвертированных значений наблюдений, и так далее.

**Среднее значение (математическое ожидание)**

*Математическое ожидание* вычисляется по формуле:

$$
 {\displaystyle {\bar {x}}={\frac {1}{n}}\sum _{i=1}^{n}x_{i}={\frac {1}{n}}(x_{1}+\ldots +x_{n})} 
 $$

In [53]:
data.Age.mean()

35.6744262295082

**Немного усложним задачу**

In [54]:
data[['Pclass', 'Age']].groupby('Pclass').mean()

Age
Pclass           
1       37.591266
2       25.266667
3       21.000000

**Медиана**

Если $x_1, x_2, ..., x_n$ – упорядоченные по возрастанию или убыванию числовые значения рассматриваемого признака, 
$n$ – объем выборки, то *медиана* - это средний элемент для нечетного  $n$ и полуcумма средних элементов для четного  $n$.

Для порядковой шкалы медиана является такой точкой на шкале, которая делит всю совокупность опрошенных на две равных части — тех, кто отметил градации меньше этой точки (либо равные ей), и тех, кто отметил градации больше этой точки.

Вопрос знатокам: можно ли посчитать медиану для категориальных не порядковых данных? 🤔

In [55]:
data.Age.median()

36.0

In [56]:
median_fare = #?
print("Медиана стоимости билета на Титанике составляла %s" % (median_fare))

SyntaxError: invalid syntax (2413915786.py, line 1)

**Мода**

Мода - значение во множестве наблюдений, которое встречается наиболее часто.

In [ ]:
data.Pclass.mode()

In [ ]:
data.Pclass.value_counts()

**Меры разброса**

Меры разброса показывают, насколько хорошо данные значения представляют совокупность. Как меры разброса применяются:

- дисперсия случайной величины и среднеквадратическое отклонение

- коэффициент вариации (это отношение среднеквадратического отклонения к среднему значению, выраженное в процентах, показывает однородность выборки)

- и так далее

**Дисперсия и среднеквадратическое отклонение**

*Дисперсия* значений признака является суммой квадратов отклонений этих значений от их среднего, деленной на число наблюдений:
$$
\sigma^2={\dfrac {\sum \limits _{i=1}^{n}\left(x_{i}-{\bar {x}}\right)}{n}}^{2}
$$
*Среднеквадратическое отклонение*, *стандартное отклонение* или *стандартный разброс* - квадратный корень из дисперсии, равный $\sigma$

Стандартное отклонение измеряется в тех же единицах, что и сама случайная величина, а дисперсия измеряется в квадратах этой единицы измерения.

In [ ]:
data[['Pclass', 'Age']].groupby('Pclass').std()

**Квантили**

Квантиль - значение, которое заданная случайная величина не превышает с фиксированной вероятностью. Если вероятность задана в процентах, то квантиль называется процентилем или перцентилем.

Например, фраза «для развитых стран 99-процентиль продолжительности жизни составляет 100 лет» означает, что ожидается, что 99 % людей проживут не более, чем 100 лет.

Относительно нашего датасета фраза "75%-перцентиль возраста пассажиров Титаника равна 38 лет" означает, что 75% пассажиров были не старше 38 лет.

Вопросы знатокам:
- Как еще можно назвать 50%-перцентиль? 🤔
- Что значит фраза "75%-перцентиль стоимости билетов на Титаник была равна 31"? 🤔

In [ ]:
data.Age.describe()

In [ ]:
data.Age.hist(bins=25)

**Описание массива нечисловых данных**

В качестве базовой описательной статистики для категориальных признаков можно использовать следующий набор характеристик:

- количество уникальных представителей массива

- частоты встречаемости этих представителей

- наиболее часто встречающиеся представители (мода распределения)

- наиболее редко встречающиеся представители

In [ ]:
data.Pclass.unique()

In [ ]:
data.Pclass.value_counts()

In [ ]:
data.Pclass.mode()

## И еще чуть-чуть! Работа с пропущенными значениями.

Вернемся к info(). Мы видим, что из 891 наблюдения у нас только 714 ненулевых значений. Значит, у этих пассажиров возраст неизвестен. Ваши идеи, что можно с этим сделать?

In [57]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 183 entries, 1 to 889
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  183 non-null    int64  
 1   Survived     183 non-null    int64  
 2   Pclass       183 non-null    int64  
 3   Name         183 non-null    object 
 4   Sex          183 non-null    object 
 5   Age          183 non-null    float64
 6   SibSp        183 non-null    int64  
 7   Parch        183 non-null    int64  
 8   Ticket       183 non-null    object 
 9   Fare         183 non-null    float64
 10  Cabin        183 non-null    object 
 11  Embarked     183 non-null    object 
 12  FamilySize   183 non-null    int64  
 13  Alone        183 non-null    int64  
 14  Title        183 non-null    object 
dtypes: float64(2), int64(7), object(6)
memory usage: 22.9+ KB


In [ ]:
data['Age'].median() # вспомним какая у нас медиана

In [ ]:
data['Age_Median'] = data['Age'].fillna(data['Age'].median())  # сохраняю результат заполнения в новую колонку

In [ ]:
data.head() # А у всех остальных - их нормальный возраст.

Или можно просто удалить пропущенные значения:

In [ ]:
data.dropna()